src/Components/test/testassets/TestServer/Startup.cs

```diff
@@ -82,16 +79,30 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
                     options.RequestCultureProviders.Add(new CookieRequestCultureProvider());
 
                     // We want the default to be en-US so that the tests for bind can work consistently.
-                    options.SetDefaultCulture("en-US"); 
+                    options.SetDefaultCulture("en-US");
                 });
 
-                app.UseRouting();
-
-                app.UseEndpoints(endpoints =>
-                {
-                    endpoints.MapBlazorHub(typeof(Index), selector: "root");
-                    endpoints.MapFallbackToClientSideBlazor<BasicTestApp.Startup>("index.html");
-                });
+                app.MapWhen(ctx => ctx.Request.Cookies.TryGetValue("__blazor_execution_mode", out var value) && value == "server",
+                    child =>
+                    {
+                        child.UseRouting();
+                        child.UseEndpoints(childEndpoints =>
+                        {
+                            childEndpoints.MapBlazorHub();
+                            childEndpoints.MapFallbackToPage("/_ServerHost");
+                        });
+                    });
+
+                app.MapWhen(ctx => !ctx.Request.Query.ContainsKey("__blazor_execution_mode"),
```

Comment:
This is really kinda gross. I think we need to have some discussion before making a change like this. I don't want to see us keep evolving our testing strategy without careful consideration.

Reply:
We need two different entry points as the server-side components can't be rendered from static html.

Tests use the hash, but unfortunately that doesn't get send to the server. I tried to swiching to a query string but that was causing other test problems. 

The cookie is the cheapest thing we can do to discriminate between server-side and client-side. The way it works is as follows.

Test opens a browser instance for the first time:
* There's no cookie, so it goes to index.html
* Index.html checks the #server hash
* If present, 
   * Removes the hash
  * Adds the cookie to the browser session.
  * Refreshes the page.
* From there on the rest of the requests go through the server endpoint.

I filed an issue with suggested improvements. Given where we are, I suggest we tackle them in the following weeks. This is not urgent and we can use the following weeks to implement a proper approach without rushing.
https://github.com/aspnet/AspNetCore/issues/13186